lnwallet/btcwallet/psbt.go

```diff
@@ -64,6 +85,164 @@ func (b *BtcWallet) FundPsbt(packet *psbt.Packet, minConfs int32,
+	// to it already and try to sign it. We do expect that we're the last
+	// ones to sign. If there is any input without witness data that we
+	// cannot sign because it's not our UTXO, this will be a hard failure.
+	tx := packet.UnsignedTx
+	sigHashes := txscript.NewTxSigHashes(tx)
+	for idx := range tx.TxIn {
+		in := packet.Inputs[idx]
+
+		// We can only sign if we have UTXO information available. Since
+		// we don't finalize, we just skip over any input that we know
+		// we can't do anything with. Since we only support signing
+		// witness inputs, we only look at the witness UTXO being set.
+		if in.WitnessUtxo == nil {
+			continue
+		}
+
+		// Skip this input if it's got final witness data attached.
+		if len(in.FinalScriptWitness) > 0 {
+			continue
+		}
+
+		// Skip this input if there is no BIP32 derivation info
+		// available.
+		if len(in.Bip32Derivation) == 0 {
+			continue
+		}
+
+		// TODO(guggero): For multisig, we'll need to find out what key
+		// to use and there should be multiple derivation paths in the
+		// BIP32 derivation field.
+
+		// Let's try and derive the key now. This method will decide if
+		// it's a BIP49/84 key for normal on-chain funds or a key of the
+		// internal 1017 key scope.
+		derivationInfo := in.Bip32Derivation[0]
+		privKey, err := b.deriveKeyByBIP32Path(derivationInfo.Bip32Path)
+		if err != nil {
+			return fmt.Errorf("error deriving signing key: %v", err)
+		}
+
+		// We need to make sure we actually derived the key that was
+		// expected to be derived.
+		pubKeysEqual := bytes.Equal(
+			derivationInfo.PubKey,
+			privKey.PubKey().SerializeCompressed(),
+		)
+		if !pubKeysEqual {
+			return fmt.Errorf("derived public key %x does not "+
+				"match bip32 derivation info public key %x",
+				privKey.PubKey().SerializeCompressed(),
+				derivationInfo.PubKey)
+		}
+
+		// Do we need to tweak anything? Single or double tweaks are
+		// sent as custom/proprietary fields in the PSBT input section.
+		privKey = maybeTweakPrivKeyPsbt(in.Unknowns, privKey)
+		pubKeyBytes := privKey.PubKey().SerializeCompressed()
+
+		var subScript []byte
+		switch {
+		// It's a NP2WKH input:
+		case len(in.RedeemScript) > 0:
+			subScript = in.RedeemScript
+
+			builder := txscript.NewScriptBuilder()
+			builder.AddData(in.RedeemScript)
+			packet.Inputs[idx].FinalScriptSig, err = builder.Script()
+			if err != nil {
+				return fmt.Errorf("error building np2wkh "+
+					"script: %v", err)
+			}
+
+		// It's a P2WSH input:
+		case len(in.WitnessScript) > 0:
+			subScript = in.WitnessScript
+
+		// It's a P2WKH input:
+		default:
+			subScript = in.WitnessUtxo.PkScript
+		}
+
+		sig, err := txscript.RawTxInWitnessSignature(
+			tx, sigHashes, idx, in.WitnessUtxo.Value, subScript,
+			in.SighashType, privKey,
+		)
+		if err != nil {
+			return fmt.Errorf("error signing input %d: %v", idx,
+				err)
+		}
+		packet.Inputs[idx].PartialSigs = append(
```

Comment:
As long as only one goroutine accesses the packet at any given time, this should be safe, yes. Or what do you mean by concurrent access?

Reply:
Yeah that's what I meant. Since the packet is just a pointer, was concerning other goroutines might modify it too. Checked the packet and it seems all the data are safe for concurrent access so I think it's ok.